packages/ckeditor5-engine/src/model/nodelist.ts

```diff
@@ -286,3 +287,21 @@ function makeOffsetsArray( nodes: Array<Node> ): Array<Node> {
 
 	return offsets;
 }
+
+/**
+ * Inserts an array into another array at a given index.
+ */
+function insertAt<T>( targetArray: Array<T>, insertArray: Array<T>, index: number ): void {
+	const originalLength = targetArray.length;
+	const insertLength = insertArray.length;
+
+	// Shift elements in the target array to make space for insertArray
+	for ( let i = originalLength - 1; i >= index; i-- ) {
+		targetArray[ i + insertLength ] = targetArray[ i ];
+	}
+
+	// Copy elements from insertArray into the target array
+	for ( let i = 0; i < insertLength; i++ ) {
+		targetArray[ index + i ] = insertArray[ i ];
+	}
```

Comment:
```suggestion
	// Shift elements in the target array to make space for insertArray.
	for ( let i = originalLength - 1; i >= index; i-- ) {
		targetArray[ i + insertLength ] = targetArray[ i ];
	}

	// Copy elements from insertArray into the target array.
	for ( let i = 0; i < insertLength; i++ ) {
		targetArray[ index + i ] = insertArray[ i ];
	}
```

Also, in the first loop, is this direction like this because this way you change the `targetArray` length only once? If so, maybe it would be better to change the length before the loop and then go from the beginning to the end (for clarity purposes, it's easier to read).